First draft of sccp documentation. (#13)
distributed-technologies/wiki: SCCP.md

<img src="images/Energinet-logo.png" width="250" style="margin-bottom: 3%">

> **Secure compute and communication platform** has emerged as a weapon to deal with some of the problems the energisystem is facing, as the energy production is becomming more renewable. One of the bigger problems are that renable energi forces the system to run close to the edge as the production of energy becomes highly flucturating. This calls for improved algorithms that helps operators of the critical systems make decitions. The SCCP is the foundation on which these models will live in any given situation, providing a platform that raises the bar for securety both for communication and compute. 

Reviewer: I think the "has emerged as a weapon" might be a bit colorful 😄.

Reviewer: Is it the energy production that is becoming more renewable or is it the energy?

Reviewer: since we write communication, should compute also be computation? I am unsure

Reviewer: ```suggestion
**Secure compute and communication platform** has emerged as a weapon to deal with some of the problems the energy system is facing, as the energy production is becoming more renewable. One of the bigger problems is that renewable energy forces the system to run close to the edge as the production of energy becomes highly fluctuating. This calls for improved algorithms that helps operators of the critical systems make decisions. The SCCP is the foundation on which these models will live in any given situation, providing a platform that raises the bar for security both for communication and computation. 
```

### Architecture

> The platform is essentially a kubernetes cluster with an agent supporting the Gitops workflow. Following the Gitops workflow all infrastructure is configured as code in seperate service git repositories. A single environment repository holds the desired configuration of each service and the agent continously pulls them in and changes the actual state accordingly.   

Reviewer: should agent be called operator instead? I think that is what it is in GitOps.

Reviewer: seperate -> separate
Also, I think it sounds weird to say service git repositories, since people might not be aware of what we mean by service.

### Hosting

> The platform is cloud agnostic, meaning that will be able to be setup both in bare metal situations as well as deployed in any cloud provider thus greatly increasing the stability and scalability of the platform, because vendor lock is avoided. Compiling our own linux operating system, we reduce the dependencies and packages to a minimum, further reducing the potential attack vectors for adversaries. Basically the only features enabled are the ones that are needed to spin up a kubernetes cluster. 

Reviewer: ```suggestion
The platform is cloud agnostic, meaning that it will be able to be deployed onto both bare metal as well as on any cloud provider thus greatly increasing the stability and scalability of the platform, because vendor lock-in is avoided. Compiling our own Linux operating system, we reduce the dependencies and packages to a minimum, further reducing the potential attack vectors for adversaries. Basically, the only features enabled are the ones that are needed to start up a Kubernetes cluster. 
```

Reviewer: I don't quite understand how being cloud agnostic increases the stability and scalability of the platform. Can we give some arguments for that?

Author: See new text.

### Security

> Taking the next step in security the platform is enabeling a feature to fuse hardware and software security in a trusted execution environment. In its essence this means that the hardware will not run any software that does not have the correct digital signatures and the software will not run on any hardware that does not have the correct digital signature.

Reviewer: ```suggestion
Taking the next step in security the platform is enabling a feature to fuse hardware and software security in a trusted execution environment. In its essence this means that the hardware will not run any software that does not have the correct digital signatures and the software will not run on any hardware that does not have the correct digital signature.
```

### Technologies 

> All technologies that is used to build the SCCP is open source.

Reviewer: ```suggestion
All technologies that are used to build the SCCP are open source.
```

#### [ArgoCD](https://github.com/argoproj/argo-cd/) 

> ArgoCD is used as the GitOps tool that contineously delivers changes to the platform based on pull request changes.

Reviewer: ```suggestion
ArgoCD is used as the GitOps tool that continuously delivers changes to the platform based on changes in the environment repositories.
```

#### [Prometheus](https://github.com/prometheus/prometheus)

> Prometheus is used as the monitoring time series database, that contineously scrabes metrics from all containers that expose them.

Reviewer: ```suggestion
Prometheus is used as the monitoring time series database, that continuously scrapes metrics from all containers that expose them.
```

#### [Grafana](https://github.com/grafana/grafana)
Grafana is used as the monitoring visualization tool, where dashboards provide insights into the metrics.

#### Ceph

> Ceph is used through an [opensource Rook operator](https://github.com/rook/rook) as a distributed persistent storage, handeling jobs like replication to avoid data loss at machine failiure. 

Reviewer: ```suggestion
Ceph is used through an [opensource Rook operator](https://github.com/rook/rook) as a distributed persistent storage, handling jobs like replication to avoid data loss at machine failure. 
```

#### [Metallb](https://github.com/metallb/metallb)

> Metallb is used as the software loadbalancer delegating ips to services that needs a static ip. 

Reviewer: ```suggestion
Metallb is used as the software loadbalancer delegating IPs to services that needs a static IP. 
```

#### [Traefik](https://github.com/traefik/traefik)

> Traefik is used an ingress controller handeling ingress events to route communication.

Reviewer: ```suggestion
Traefik is used as an ingress controller handling ingress events to route communication.
```

### Roadmap
- [x] Monitoring (done)
- [x] Ingress (done)
- [x] Loadbalancer (done)
- [x] Distributed storage (done)

> - [ ] TTE

Reviewer 2: TEE?

- [ ] PKI infrastructure

> - [ ] One click build pipeline

Reviewer: Is this a complete roadmap?

Author: This is just meant as an example of what we could but in the docs to reflect the status of the platform, and it should not be seen as an exhaustive list of features. If we choose to keep it, we will need to take some high-level features from the RTP and put in here.

<img src="images/Energinet-logo.png" width="250" style="margin-bottom: 3%">

> **Secure compute and communication platform** has emerged as a tool to deal with some of the problems the energy system is facing, as the production is shifting towards more renewable energy. One of the bigger problems is that renewable energy forces the system to run close to the edge as the production of energy becomes highly fluctuating. This calls for improved algorithms that help operators of critical systems make decisions. The SCCP is the foundation on which these models will live in any given situation, providing a platform that raises the bar for security both for communication and computation. 

Reviewer 2: I think we should look at it more broadly, we made the SCCP platform to be able to define a secure environment in code. The environment are required to be strictly controlled with the help of TEE and signed code, which ensures only signed and allowed code are executed on the platform.

Renewable energy itself is not the reason alone why we are doing this.

Author: New text.

### Architecture

> The platform is essentially a Kubernetes cluster with an operator supporting the Gitops workflow. Following the Gitops workflow, all infrastructure is configured as code in separate git repositories as individual services. A single environment repository holds the desired configuration of each service and the agent continuously pulls them in and changes the actual state accordingly.   

Reviewer 2: "GitOps" ..

The environment itself should be defined based on a single repository in the near future, describing the infrastructure provisioning, enabled system services, and referring hosted applications environment repositories.

I think this should be made more clear.

The platform is cloud-agnostic, meaning that it will be able to be deployed both in bare metal situations as well as on deployed in any cloud provider thus greatly increasing the stability and scalability of the platform because vendor lock-in is avoided. Avoiding vendor lock-in we reduce the risk caused by cloud providers having an outage, a security event, or simply discontinues services, as a cloud-agnostic platform can seamlessly move to wherever fits its demands. 
Compiling our own Linux operating system, we reduce the dependencies and packages to a minimum, further reducing the potential attack vectors for adversaries. Basically, the only features enabled are the ones that are needed to start up a Kubernetes cluster. 

### Security
Taking the next step in security, the platform is enabling a feature to fuse hardware and software security in a trusted execution environment. In its essence, this means that the hardware will not run any software that does not have the correct digital signatures and the software will not run on any hardware that does not have the correct digital signature.

### Technologies 
All technologies that are used to build the SCCP are open source.

> #### [ArgoCD](https://github.com/argoproj/argo-cd/) 

Reviewer 2: I don't think this list is essential here. 
I would suggest moving it to a seperate file.
More over we should add a notice describing the licenses for our dependencies, similar to DataHub
https://github.com/Energinet-DataHub/green-energy-hub/blob/main/NOTICE.md

Author: Since we do not have a single environment repository yet, should we make a new task on the NOTICE docs?

Author: I have added a draft of the notice. We need to input email and dependencies.

<img src="images/Energinet-logo.png" width="250" style="margin-bottom: 3%">

**Secure compute and communication platform** 
The energy system is the most critical infrastructure needed to uphold basic functions in society, and it is key that it can not be subject to any adversarial attacks. The **SCCP** fuses hardware and software security in a **Trusted Execution Environment** to make sure that everything that is executed has been signed and approved to 
run on the platform and only on the platform, meaning that a signed application will refuse to run anywhere else.  

> 

Reviewer 2: Agreed with both points.

### Architecture
The platform is essentially a Kubernetes cluster with an operator supporting the GitOps workflow. Following the GitOps workflow, all infrastructure is configured as code in separate git repositories as individual services. The environment itself should be defined based on a single repository in the near future, describing the infrastructure provisioning, enabled system services, and referring hosted applications environment repositories.

### Hosting

> The platform is cloud-agnostic, meaning that it will be able to be deployed both in bare metal situations as well as on deployed in any cloud provider thus greatly increasing the stability and scalability of the platform because vendor lock-in is avoided. Avoiding vendor lock-in we reduce the risk caused by cloud providers having an outage, a security event, or simply discontinues services, as a cloud-agnostic platform can seamlessly move to wherever fits its demands. 

Reviewer 3: I do not see how being cloud-agnostic makes the platform more scalable. Deploying in the cloud at multiple availability zones provides the same "outage"-protection. What is meant by "security event"?

Compiling our own Linux operating system, we reduce the dependencies and packages to a minimum, further reducing the potential attack vectors for adversaries. Basically, the only features enabled are the ones that are needed to start up a Kubernetes cluster. 

### Security

> Taking the next step in security, the platform is enabling a feature to fuse hardware and software security in a trusted execution environment. In its essence, this means that the hardware will not run any software that does not have the correct digital signatures and the software will not run on any hardware that does not have the correct digital signature.

Reviewer 3: In essense i think we should hold of on putting these claims into writing untill they are developed.

Reviewer 2: But we could still tell what the goal is, as long as it is clear that it is not yet fully developed.
